Fixes #29992 - pass stopInterval callback to handleError in API action

Ron-Lavi opened the pull request:
Whenever an API polling encounters an error,
`stopInterval` for that specific key will be passed in the `handleError` callback,
so the consumer could stop it from the action, see how I implemented it in the notification in this PR.
<!---

Thank you for contributing to The Foreman, please read the
[following guide](https://www.theforeman.org/contribute.html), in short:

* [Create an issue](https://projects.theforeman.org/projects/foreman/issues)
* Reference the issue via `Fixes #1234` in the commit message
* Prefer present-tense, imperative-style commit messages
* Mark all strings for translation, see [1]
* Suggest prerequisites for testing and testing scenarios following example above.
* Prepend `[WIP]` for work in progress to prevent bots from triggering actions
* Be patient, we will do our best to take a look as soon as we can
* Explain the purpose of the PR, attach screenshots if applicable
* List all prerequisites for testing (e.g. VMware cluster, two smart proxies...)
* Reviewers often use extensive list of items to check, have a look prior submitting [2]
* Be nice and respectful

1: https://projects.theforeman.org/projects/foreman/wiki/Translating#Translating-for-developers
2: https://github.com/theforeman/foreman/blob/develop/developer_docs/pr_review.asciidoc
-->

amirfefer commented on webpack/assets/javascripts/react_app/redux/API/APIRequest.js:
```diff
@@ -21,6 +33,11 @@ export const get = async (
     });
   } catch (error) {
+    const doesIntervalExist =
+      interval !== undefined && selectDoesIntervalExist(getState(), key);
```
nit: the undefind comparing can be removed
```js
const doesIntervalExist interval && selectDoesIntervalExist(getState(), key);
```
can we just use the selector? if the key doesn't exist it'll return false anyway, right?

Ron-Lavi replied:
thanks I'll remove the comparing, though I prefer still to check if the `action.interval` exist,
to understand if this action is using the interval middleware.. if it's not, there is a small chance that there is another action which is using the interval with the same key as the API key, and we could accidently delete it, so I prefer to be more certain.

ekohl reviewed (commented):
Is the user ever notified about this? How does the user recover? Let's say my wifi goes down for a moment. This results in a failed poll. As I read this now, the user needs to reload to recover.

On a related note, https://developer.mozilla.org/en-US/docs/Web/API/NavigatorOnLine/Online_and_offline_events could mitigate this, but it's not really guaranteed. You may be online, but need a VPN that's offline etc.

Ron-Lavi commented:
> Is the user ever notified about this? How does the user recover?

In the notification component the error is visible, but you are right, there can be some component that doesn't show any difference.

we can leave it for the user to decide whenever to destroy the interval by triggering `dispatch(stopInterval(key))` with the callbacks #7337 
or add a special key to the `action`, e.g: `clearOnError` as a shorter solution

ShimShtein commented:
We have discussed an option to pass a `cancelInterval` function into API callback, so the user will be able to decide whether to leave the interval as it is, or cancel it.

amirfefer commented:
@ShimShtein what would be the default behavior?

ShimShtein commented:
@amirfefer, what do you mean by default behaviour?
Assume code:
``` js
const myComponentsApiCallFailed = (error, cancelInterval) => {
  //do some stuff, look at the error
  if (isFatal(error)) {
    cancelInterval();
  };
};

//Then somewhere in the code we will call the API:
callApiGet('my/important/endpoint', myComponentsApiCallFailed);
```

Everything is user defined. If the user does not call `cancelInterval`, he can either change the interval, or leave the call as it is.

ronlavi2412 committed:
Fixes #29992 - pass stopInterval in handleError

Ron-Lavi commented:
rebased and updated, now `stopInterval` for that specific key will be passed in the `handleError` callback,
so the consumer could stop it from the action, see how I implemented it in the notification component in this PR.

Ron-Lavi commented on webpack/assets/javascripts/react_app/components/notifications/index.js:
```diff
@@ -32,17 +31,6 @@ class notificationContainer extends React.Component {
-    const { error, stopNotificationsPolling } = this.props;
-    if (error) {
-      const { response: { status } = {} } = error;
-      stopNotificationsPolling();
```
this logic now lives in the actions file.

ShimShtein commented on webpack/assets/javascripts/react_app/redux/actions/notifications/index.js:
```diff
@@ -18,10 +19,22 @@ import { DEFAULT_INTERVAL } from './constants';
 
-const getNotifications = url => get({ key: NOTIFICATIONS, url });
+const handleNotificationPollingError = (error, stopNotificationPolling) => {
+  stopNotificationPolling();
```
This will mean notifications will stop after one unsuccessful poll. Is it intended? Basically @ekohl's [comment](https://github.com/theforeman/foreman/pull/7718#pullrequestreview-423652334) is still valid for notifications.

Ron-Lavi replied:
thanks, moved it into the `if` context

Ron-Lavi commented:
rebased

MariaAga commented on webpack/assets/javascripts/react_app/redux/actions/notifications/index.js:
```diff
@@ -18,10 +19,22 @@ import { DEFAULT_INTERVAL } from './constants';
 
-const getNotifications = url => get({ key: NOTIFICATIONS, url });
+const handleNotificationPollingError = (error, stopNotificationPolling) => {
+  if (error?.response?.status === 401) {
```
This is a new feature that we don't use in Foreman (https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining)
And it's only available from March 2020, are we sure we want drop support for "old browsers"?

Ron-Lavi replied:
thanks for bringing this up, I think that with babel we get the polyfills for older browser
cc @jeremylenz @sharvit thoughts on this?

Ron-Lavi replied:
I think this is the supported browser list, [chrome > 26, firefox > 4...](https://github.com/zloirock/core-js#supported-engines)

MariaAga replied:
I think we need to add [https://babeljs.io/docs/en/babel-plugin-proposal-optional-chaining](https://babeljs.io/docs/en/babel-plugin-proposal-optional-chaining) to the babel plugins config in order for it to work

MariaAga replied:
I see that `@babel/plugin-proposal-optional-chaining` is installed for me in foreman so it should work, but I'm not sure where from it's installed

Ron-Lavi commented:
This PR also solves the [notification interval endless loop bug](https://github.com/theforeman/foreman/pull/7771#issuecomment-660658122),
as it uses the [handleError](https://github.com/theforeman/foreman/pull/7718/files#diff-2abb6d1defb26fc866851a507eef315cR34) API callback, instead of the `componentDidUpdate` which seem to cause the loop.

ShimShtein commented on webpack/assets/javascripts/react_app/redux/actions/notifications/index.js:
```diff
@@ -18,10 +19,22 @@ import { DEFAULT_INTERVAL } from './constants';
 
-const getNotifications = url => get({ key: NOTIFICATIONS, url });
+const handleNotificationPollingError = (error, stopNotificationPolling) => {
+  if (error?.response?.status === 401) {
```
Nitpick: `error` will always have a value
```suggestion
  if (error.response?.status === 401) {
```

Ron-Lavi replied:
true, thanks :+1:

tbrisker reviewed (approved):
Thanks @laviro, tested, this does resolve the loop.
